### ISSUE TITLE
add require-global option to composer.json

### ISSUE THREAD

**Issue by schmunk42** [composer/composer#3607]
Since you can install packages globally, you should also be able to specify a global requirement in a composer.json file of a project.

See also https://github.com/composer/composer/pull/3082 - the plugin mentioned there for example requires to be installed globally to work correctly, if you do not have a global installation of the plugin you're getting no or non-obvoius error-messages and the package installation seems to work fine, but the application as a whole won't work as expected.

Proposed syntax in `composer.json`:

```
"require-global": {
  "fxp/composer-asset-plugin": "~1.0"
}
```

**Comment by francoispluchino** [comment]
@schmunk42 I had already making a proof of concept for this feature, and if @Seldaek and @naderman are interested, in this case, I will create a new PR.

But personally, I think it would be better not to use the standard behavior of Composer, and extract asset dependencies to the "extra" section. In this way, we can correct directly in the plugin, the problem of camelcase of the package names, together with the opportunity to install the plug-in project mode. In addition, we can add a very interesting feature of Bower: possible to select a particular version of a dependency, if there is a version conflict (a new issue will be created in the plugin repository for this feature).

**Comment by bishopb** [comment]
I presume that [platform packages](https://getcomposer.org/doc/02-libraries.md#platform-packages) would be supported in this new block?  Like:
```json
"require-global": {
    "php": "~5.5",
    "fxp/composer-asset-plugin": "~1.0@dev"
}
```

On a side note, composer-asset-plugin from #3082 doesn't need to be installed globally: it just needs to be installed first.

**Comment by francoispluchino** [comment]
@bishopb For the plugin, your proposal is not functional in project mode, because Composer will not find any package for assets, simply because the plugin is not installed yet. and Composer will not install the plugin, because there are errors (packages not found).

The PR #3082 allows precisely to allow to the plugins to be installed, and restart analysis where appropriate (if the plugin asks  specifically a new analysis).

**Comment by bishopb** [comment]
@francoispluchino I hear what you're saying.  Let me clarify what I meant about "first":

1. create a `composer.json` containing `"require": { "fxp/composer-asset-plugin": "~1.0@dev" }`
1. run `composer --install`
1. update `composer.json` containing `"require": { "fxp/composer-asset-plugin": "~1.0@dev", "npm-asset/jquery": "~2.1" }`
1. run `composer --install`

I realize that is non-standard and non-repeatable usage.  It does work, however.  And it underscores my point: the plugin needs to come *first*.  Global installation is *one* way to get first, but not the only.

**Comment by francoispluchino** [comment]
Yes, but in this way, it works a single time. Keep your latest version of composer.json file, and launch an installation (without vendor, and without composer.lock), Composer will throw an exception "package not found".

Also, if you add a dependency that requires the plugin (because it also uses an asset dependency), Composer will throw an exception "package not found", because the plugin is not installed yet. It must not be forgotten that the asset dependencies are not placed only in the root Composer package.

**Comment by schmunk42** [comment]
Example for what can go wrong, https://github.com/dmstr/yii2-adminlte-asset/issues/10#issuecomment-69298944

**Comment by francoispluchino** [comment]
@schmunk42 I'll move the "require" section in "extra.asset-require" section, and maintain backward compatibility. A issue will be created soon to discuss it. This will solve several problems at the same time, whose the installation in the project mode and the problem for the camelcase for Bower.

**Comment by francoispluchino** [comment]
To see the first version of this proposal, this the [comment](https://github.com/composer/composer/pull/3082#issuecomment-56285632)

**Comment by schmunk42** [comment]
> "require" section in "extra.asset-require"

Sounds like a very good solution!

**Comment by curry684** [comment]
Isn't this a reverse solution?

If a plugin can only work if installed globally, it should *itself* check whether it's installed globally, and otherwise bug out with a big fat error. It makes no sense for the dependents to have to declare that, that's the plugin author's responsibility.

So no I don't see merits to a `require-global` section, if the true issue seems to be that a plugin can't detect (reliably) whether it's installed globally or not.

**Comment by schmunk42** [comment]
It's about the case that the plugin is not installed at all.

@francoispluchino Would it be an option, that we add `fxp/composer-asset-plugin` to `require` and the plugin does the check for global installation like @curry684 mentioned?

**Comment by cebe** [comment]
@curry684 adding the plugin to `require` section will break composer so the plugin is not there in the first place. It can not do any checks until it is installed.

**Comment by cebe** [comment]
@schmunk42 composer will only install it if all dependencies are found. that is not the case if there are requirements for bower/npm packages so nothing will be installed.

**Comment by schmunk42** [comment]
@cebe Thank you. I remember the long discussions about this hen-and-egg problem now :)

**Comment by curry684** [comment]
I understand the issue for this specific plugin, I just don't agree with the proposed solution as it's an invalid reversal of responsibility. It's not like the resolver can meaningfully implement it, as it would imply knowing about needing the plugin through `require-global` before it can know from the main dependencies that the package itself is ever going to be selected in the next phase of resolution. So `require-global` cannot be implemented in a way that solves this problem, it just moves the hen-and-egg issue.

I think right now the only real solution is educating users that they need to install that plugin before using dependent packages. The only structural solution I can think of would implement namespaces that are ignored unless a proper handler is installed.

**Comment by Seldaek** [comment]
require-global could be implemented as a simple check for presence of that package in that version in the global deps. If present then ok otherwise error. *If* we implement it it wouldn't be a part of the require/require-dev dependency resolution at all IMO.

**Comment by curry684** [comment]
In that case it would also have to be called something like `"uses"` instead to loud and clear signify that it isn't part of the dependency tree, and will never ever cause anything to be installed automatically, just giving a clearer error that a plugin isn't installed which is required by one of the selected packages.

Also it would have to be *crystal clear* that this is a root-only section then as it can never work in nested dependencies.

Imo the underlying problem is just that `fxp/composer-asset-plugin` never should've mingled its own config with the `require` sections. If it had just used `extra/assets` config section instead all of these issues would've been academic.

**Comment by naderman** [comment]
I agree with @curry684 that really a plugin ought to not have to be installed globally prior to running composer to begin with. So far global plugins are an option, but entirely avoidable, adding this feature we would advocate in favour of people installing global plugins, so I really don't think this should be implemented.

**Comment by cebe** [comment]
> Imo the underlying problem is just that fxp/composer-asset-plugin never should've mingled its own config with the require sections. If it had just used extra/assets config section instead all of these issues would've been academic.

It can not just use an extra section. It has to insert packages into `require` to ensure they are part of the dependency resolving process.

> Also it would have to be crystal clear that this is a root-only section then as it can never work in nested dependencies.

Sure, it has to be checked before starting to resolve dependencies because the plugin is needed for the dependency resolving step.

**Comment by francoispluchino** [comment]
@naderman, @curry684, I agree with you. I really don't think this should be implemented. However a less elegant solution will achieve this functionality:

Create a new plugin to use in a project mode that would validate the presence of the plugin in global mode.

Concerning the `fxp/composer-asset-plugin`, extract the resolution of the asset dependencies is impossible, because the plugin adds only the VCS Repositories.

**Comment by francoispluchino** [comment]
I remember to an old debate about this issue, and taking into consideration the [last comment](https://github.com/composer/composer/issues/3607#issuecomment-187668571) of @naderman, I wonder whether it is not better to have a dedicated section of the plugins required in the project (instead of `require-global`) and only available to the root Composer package.

**Example:**
```json
{
    "require-plugin": {
        "fxp/composer-asset-plugin": "~1.0"
    }
}
```

In this case, Composer installs the plugins, and resolves dependencies after installing plugins (using the plugins, of course).

Indeed, there is already a system for installing the plugins directly in the `require` section, but it does not solve the problem for plugins adding functionality to Composer. We can lock this feature for plugins using the next API only (2.0).

This is a much more elegant solution than PR #3082, which does the same job, but that looks like the proposed "require-global".

Between my 3 proposals (restart the solver(see the PR), require-global, and require-plugin), the solution that does not bother anyone, and allowing to remain in the recommendations, is to add a section `require-plugin` (or another name) and not `require-global`.

I think this is a good compromise.

**Comment by Soullivaneuh** [comment]
> require-global could be implemented as a simple check for presence of that package in that version in the global deps. If present then ok otherwise error.

In this case, a `--ignore-global-reqs` working like actual `--ignore-platform-reqs` option would be great.

**Comment by lwrbakro** [comment]
+1

**Comment by brunoric** [comment]
+1

**Comment by Seldaek** [comment]
Closing as there is little chance this'll ever happen sorry

**Comment by schmunk42** [comment]
I am about to add this to my **projects**:

```
    "conflict": {
        "fxp/composer-asset-plugin": "*"
    },
```

Which prevents installation of the plugin in **project scope**, you be nice to get some feedback, if there might be any problems with this approach.
